src/hypothesis/internal/conjecture/utils.py

```diff
@@ -225,66 +231,97 @@ def biased_coin(data, p):
-            if acceptance > 0:
-                self.extras.append(i)
-        self.extras.sort(key=self.acceptance.__getitem__, reverse=True)
 
-    def sample(self, data):
-        while True:
-            data.start_example()
-            # We always draw the acceptance data even if we don't need it,
-            # because that way we keep the amount of data we use stable.
-            i = integer_range(data, 0, len(self.table) + len(self.extras) - 1)
-            if i < len(self.table):
-                result = self.table[i]
-                data.stop_example()
-                return result
+        self.table = [[i, None, None] for i in hrange(n)]
+
+        total = sum(weights)
+
+        num_type = type(total)
+
+        zero = num_type(0)
+        one = num_type(1)
+
+        small = []
+        large = []
+
+        probabilities = [w / total for w in weights]
+        scaled_probabilities = []
+
+        for i, p in enumerate(probabilities):
+            scaled = p * n
+            scaled_probabilities.append(scaled)
+            if scaled == 1:
+                self.table[i][2] = zero
+            elif scaled < 1:
+                small.append(i)
+            else:
+                large.append(i)
+        heapq.heapify(small)
+        heapq.heapify(large)
+
+        while small and large:
+            lo = heapq.heappop(small)
+            hi = heapq.heappop(large)
+
+            assert lo != hi
+            assert scaled_probabilities[hi] > one
+            assert self.table[lo][1] is None
+            self.table[lo][1] = hi
+            self.table[lo][2] = one - scaled_probabilities[lo]
+            scaled_probabilities[hi] = (
+                scaled_probabilities[hi] + scaled_probabilities[lo]) - one
+
+            if scaled_probabilities[hi] < 1:
+                heapq.heappush(small, hi)
+            elif scaled_probabilities[hi] == 1:
+                self.table[hi][2] = zero
             else:
-                result = self.extras[i - len(self.table)]
-                accept = not biased_coin(data, 1 - self.acceptance[result])
-                data.stop_example()
-                if accept:
-                    return result
+                heapq.heappush(large, hi)
+        while large:
+            self.table[large.pop()][2] = zero
+        while small:
+            self.table[small.pop()][2] = zero
+
+        for entry in self.table:
+            assert entry[2] is not None
+            if entry[1] is None:
+                entry[1] = entry[0]
+            elif entry[1] < entry[0]:
+                entry[0], entry[1] = entry[1], entry[0]
+                entry[2] = one - entry[2]
+        self.table.sort()
+
+    def sample(self, data):
+        data.start_example()
+        i = integer_range(data, 0, len(self.table) - 1)
+        base, alternate, alternate_chance = self.table[i]
+        use_alternate = biased_coin(data, alternate_chance)
+        data.stop_example()
+        if use_alternate:
+            return base
```

Comment:
Surely if `use_alternate` is True, we should return `alternate`, not `base`?

Reply:
😳 yes. Yes we should.

Fixed and added a test that the actual probability distribution is compatible with the desired one.